module_utils/nxos/facts/base.py

```diff
@@ -0,0 +1,95 @@
+import re
+
+from copy import deepcopy
+
+from ansible.module_utils.six import iteritems
+
+
+class FactsBase(object):
+
+    generated_spec = {}
+    ansible_facts = {'net_configuration': {}}
+
+    def __init__(self, argspec, subspec=None, options=None):
+        spec = deepcopy(argspec)
+        if subspec:
+            if options:
+                facts_argument_spec = spec[subspec][options]
+            else:
+                facts_argument_spec = spec[subspec]
+        else:
+            facts_argument_spec = spec
+
+        self.generated_spec = self.generate_dict(facts_argument_spec)
+
+    def generate_dict(self, spec):
+        """
+        Generate dictionary which is in sync with argspec
+
+        :param spec: A dictionary which the argspec of module
+        :rtype: A dictionary
+        :returns: A dictionary in sync with argspec with default value
+        """
+        obj = {}
+        if not spec:
+            return obj
+
+        for k, v in iteritems(spec):
+            if 'default' in v:
+                d = {k: v['default']}
+            else:
+                d = {k: None}
+            obj.update(d)
+
+        return obj
+
+    def parse_conf_arg(self, cfg, arg):
+        """
+        Parse config based on argument
+
+        :param cfg: A text string which is a line of configuration.
+        :param arg: A text string which is to be matched.
+        :rtype: A text string
+        :returns: A text string if match is found
+        """
+        match = re.search(r'%s (.+)(\n|$)' % arg, cfg, re.M)
```

Comment:
Can we make this more generic by making this accept the pattern instead of hardcoding it? For e.g., something like [this](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/network/frr/frr_facts.py#L143) which is being utilized [here](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/network/frr/frr_facts.py#L187-L208).

Reply:
@NilashishC this is base class ***only*** for NXOS facts. This method is not a generic method for all kind of regex matches, rather for strings that start with the key we are looking for the value right after it.
For example,
`description Configured by Ansible`
`feature nxapi`
`hostname test`

The example you shared has the exact same regex pattern for all the keys, so the regex pattern can act as a generic pattern for all those keys imo.
Maybe we need a better method name?